File: MATLAB/Method/Statistical_methods/Frequency_domain/SDCC/SDCC.m
function [Pulse_shape_discrimination_factor] = SDCC(Pulse_signal, ROI_left_endpoint)

%--------------------------------------------------------------------------
% Performs pulse shape discrimination using Simplified Digital Charge 
% Collection (SDCC).
% 
% Calculates discrimination factor as the log of squared sum over a region
% of interest (ROI) in the pulse tail, where neutron and gamma signals
% show maximum decay rate differences.
%
% Input:
%   ROI_left_endpoint: Starting point of the ROI in the pulse tail
%
% Reference:
% Shippen, David I., Malcolm J. Joyce, and Michael D. Aspinall. 
% "A wavelet packet transform inspired method of neutron-gamma discrimination."
% IEEE Transactions on Nuclear Science 57.5 (2010): 2617-2624. 325-332.
%--------------------------------------------------------------------------
    
% Set default ROI start point
if nargin < 2
    ROI_left_endpoint = 70;  % Default position in pulse tail
end

% Initialize output array
Num_signal = size(Pulse_signal, 1);
Pulse_shape_discrimination_factor = zeros(Num_signal, 1);

% Process signals in parallel
parfor i = 1:Num_signal
    Pulse_signal_vector = Pulse_signal(i, :);

    % Extract region of interest from pulse tail
    ROI = Pulse_signal_vector(ROI_left_endpoint:end);

    % Calculate discrimination factor:
    % log(sum(ROI^2))
    Pulse_shape_discrimination_factor(i) = log(sum(ROI.^2));
end

end

File: Method/Statistical_methods/Frequency_domain/sdcc.py
"""
Simplified Digital Charge Collection (SDCC) method for pulse shape discrimination.

This module implements a simplified domain transformation approach to pulse shape
discrimination by analyzing the decay rate differences between class 1 and class 2
signals in a specific region of interest (ROI).

Reference:
- Shippen, David I., Malcolm J. Joyce, and Michael D. Aspinall.
  "A wavelet packet transform inspired method of neutron-gamma discrimination."
  IEEE Transactions on Nuclear Science 57, no. 5 (2010): 2617-2624.
"""

import numpy as np
from typing import Union


def get_psd_factor(pulse_signal: np.ndarray, ROI_left_endpoint: Union[int, None] = 70) -> np.ndarray:
    """
    Performs pulse shape discrimination based on Simplified Digital Charge Collection (SDCC).

    SDCC computes a simplified domain transformation to serve as the discrimination factor.
    For each pulse signal, the region of interest (ROI) is defined starting from the provided
    ROI_left_endpoint to the end of the signal. The discrimination factor is computed as the
    natural logarithm of the sum of the squared values of the ROI.

    Args:
        pulse_signal: 2D numpy array where each row represents a pulse signal
        ROI_left_endpoint: The starting index (0-based) for the region of interest where the decay
                          rate difference between class 1 and class 2 signals is the greatest (default: 70)

    Returns:
        A numpy array of discrimination factors for each pulse signal
    """
    num_signals = pulse_signal.shape[0]
    discrimination_factors = np.zeros(num_signals)

    for i in range(num_signals):
        # Extract the pulse signal vector for the current pulse
        signal = pulse_signal[i, :]
        # Define the region of interest (ROI) starting from ROI_left_endpoint to the end
        ROI = signal[ROI_left_endpoint:]
        # Compute the discrimination factor as the logarithm of the sum of squared ROI values
        discrimination_factors[i] = np.log(np.sum(ROI ** 2))

    return discrimination_factors

